### PR TITLE
Add Support for Album Cover Art Files

### LINKED ISSUE #149: Include album art file in collections
## Cover art file

We want to store album cover art in an external collection. (See #43 for the original request.)

Here’s the suggested documentation of the feature.

> * **`albumart_file`** When `true`, writes album artwork as separate image files alongside the tracks in each album directory. The artwork file  is placed in the same directory as the first track of the album, using the  filename configured by the [`art_filename`](https://beets.readthedocs.io/en/stable/reference/config.html#art-filename)   setting. Note that artwork files are not automatically  removed when albums are deleted from the collection.

## Test cases

* Album art is copied/linked for album added to a collection
* Album art is updated for album already in collection
* Album art is not added for album that does not have cover art
* Album art is not added for single tracks where album is not matched
* Album art copy works even if albums share the same directory

## Implementation notes

* Align with behavior of `fetchart` and `embedart` plugins
* Use `Album.artpath` to locate art and `Album.art_destination` to determine location, See `fetchart.Plugin._set_art`

### PR DESCRIPTION
**Pull request by maxaudron**
As this was a personal show-stopper for me, I've gone and implemented at least basic support for copying album cover art over. For my use-case, I keep an opus transcode in sync for mobile devices with limited storage, but still want reasonably decent size album cover art. Embedding a 1MB file into each 3MB track is a rather big waste of space, so having a `cover.png` makes the most sense to me.

I'm using the convert plugins [`copy_album_art`](https://github.com/beetbox/beets/blob/c1877b7cf5371f5399fbf80f5e9e64bd119b9917/beetsplug/convert.py#L486) fn and supports resizing with the convert settings.

### Test Cases

- [X] Album art is copied/linked for album added to a collection
- [X] Album art is updated for album already in collection
- [X] Album art is not added for album that does not have cover art
- [ ] Album art is not added for single tracks where album is not matched

This cannot be tested for afaik and should never be the case as the item.art_destination() fn doesn't exist
For the future it might make sense to expand art embedding to optionally only apply to singles?

- [ ] Album art copy works even if albums share the same directory

I'm not sure what you mean by this, like Multi Disc releases without disc subfolders but each their own artwork?
I don't think this would work at all, convert standardizes all art files to one name e.g. "cover.png" by default
as that is also what players will read. Afaik there is no player support to read the art path from a tag and one of a few "standard" filenames are just picked from.

- [X] Album art is copied if track matches, but not the album

### ~~ToDo~~

~~I've not added support to `SymlinkView` for now. The `copy_album_art()` impl from the convert plugin always copies the art file if a resize is enabled, which I don't think is wished behavior for symlinking here. In the same vein, the options are somewhat inconsistent with the enable for `copy_album_art` being pulled from convert, and the `albumart_maxwidth` for copy from convert, but for embed from alternatives config.~~

~~If you want to i'd be glad to spend some additional time pulling the copy impl and options into alternative and unify this with #60 for embed.~~

closes #149
supersedes #58

### CODE REVIEW

**Comment by maxaudron** [comment]
Realized I needed reformatting of the album art just for a collection anyways, so here is the whole impl pulled into alternatives and I unified the whole thing with `embed`.

So both options now support image resizing, reformatting and disabling of interlacing and are configured on a per collection basis.

**Comment by endcredits33** [comment]
One thing I've noticed if copying is enabled - if an album already in the collection gets a tag update that results in a folder name change, the old folder stays with just the cover art file in it.

File: test/helper.py [hunk 1]
```diff
@@ -267,6 +278,14 @@ def add_external_album(self, ext_name: str, **kwargs: str):
     def get_path(self, item: Item, path_key: str = "alt.myexternal") -> Path:
         return Path(item[path_key])
 
+    def get_album_path(self, album: Album, path_key: str = "alt.myexternal") -> Path:
+        item = album.items().get()
+        if item:
+            head, _ = os.path.split(self.get_path(item, path_key=path_key))
+            return head
```
Comment (geigerzaehler) [inline test/helper.py]:
```suggestion
            return self.get_path(item, path_key=path_key).parent
```

File: test/cli_test.py [hunk 2]
```diff
@@ -350,6 +371,175 @@ def test_unkown_collection(self):
...
+                "album_art_copy": False,
+                "album_art_maxwidth": None,
+                "album_art_format": None,
+                "album_art_deinterlace": False,
+                "album_art_quality": 0,
+            }
+        }
+        self.external_config = self.config["alternatives"]["myexternal"]
+
+    def touch_art(self, artpath: bytes, image_path: Path):
```
Comment (geigerzaehler) [inline test/cli_test.py]: Please keep this as a top-level function. I would like to avoid methods that don’t access the instance.

File: test/cli_test.py [hunk 3]
```diff
@@ -350,6 +371,175 @@ def test_unkown_collection(self):
...
+        self.runcli("alt", "update", "myexternal")
+
+        dest_path = Path(str(dest, "utf8"))
+        mtime_1 = dest_path.stat().st_mtime
+        self.runcli("alt", "update", "myexternal")
+        mtime_2 = dest_path.stat().st_mtime
+
+        assert mtime_1 == mtime_2
+
+    def test_copy_art(self, tmp_path: Path):
```
Comment (geigerzaehler) [inline test/cli_test.py]: It would be great if we had a test case that checks that the art is only reprocessed and copied if the timestamp is newer than the existing art timestamp.

**Comment by geigerzaehler** [comment]
> * [ ]  Album art copy works even if albums share the same directory
> 
> 
> I'm not sure what you mean by this, like Multi Disc releases without disc subfolders but each their own artwork? I don't think this would work at all, convert standardizes all art files to one name e.g. "cover.png" by default as that is also what players will read. Afaik there is no player support to read the art path from a tag and one of a few "standard" filenames are just picked from.

I was thinking of a directory layout like the following
```
external-dir/
  2012 /
    artist 1 - track 1.mp3
    artist 2 - track 2.mp3
```

But I think this is an edge case for cover art usage and fine if we don’t have a test for it.


> * [ ]  Album art is not added for single tracks where album is not matched
> 
> 
> This cannot be tested for afaik and should never be the case as the item.art_destination() fn doesn't exist For the future it might make sense to expand art embedding to optionally only apply to singles?

I meant that we shouldn’t copy cover art if only one track of an album is matched. But after revisiting this, I don’t think this behavior is a good idea. Users probably expect the cover art to be copied.

File: README.md [hunk 4]
```diff
@@ -271,6 +266,33 @@ following settings.
   **`formats`** is `link`, it sets the type of links to create. For
   differences between link types and examples see [Symlink Views](#symlink-views).
 
+* **`album_art_embed`** Embed album art into the media file. Default `yes`
+
+* **`album_art_copy`** Copy album art files into the collection. 
+  If `formats: link` is used then album art is linked instead.
+
+* **`album_art_maxwidth`** Downscale the embedded album art to a width
```
[the remark this answers is not in the document]
Comment (maxaudron) [reply README.md]: I started out with using the fetchart config, but found it more practical to have it per collection.
My main archive I want the full res artwork I can get, but on the ipod a quite small jpeg is plenty.

**Comment by maxaudron** [comment]
Think I should have addressed all topics now so ready for re-review. I've split up into commits per topic, with notes for each.

I am sadly not able to use uv fully as for the self-inflicted pain that is nixos, but I have the tooling all setup locally and it should hopefully be all good now.

The `ResourceWarning: unclosed file` exceptions that were in the CI previously are a bit mysterious to me, as I do not get them when running pytest locally, but did get them when building the nix package. As those drill down into the artresizer module of beets I don't think we can do anything about them and have to ignore them.

File: beetsplug/alternatives.py [hunk 5]
```diff
@@ -410,6 +488,56 @@ def finalize_converted_item(item: Item, dest: Path):
...
+            if self._config.album_art_copy:
+                self.update_art()
+
+    def update_art(self, link: bool = False):
+        for album in self.lib.albums():
+            if (
+                self._config.query.match(album)
+                or any(self._config.query.match(item) for item in album.items())
+            ) and (paths := self.get_album_art_update(album)) is not None:
+                artpath = paths[0]
```
Comment (geigerzaehler) [inline beetsplug/alternatives.py]: This is already more explicit and has less indirection. But I would like to make it as clean as possible. The main problem is that there are about five conditions that need to be satisfied for the art to be copied. Currently the code for checking the conditions is spread between here and `self.get_album_art_update()`. I would like to have them all in one place so they are easier to understand. And to make the checks more readable and avoid deep nesting I’d suggest the following format:

```python
for album in self.lib.albums():
    if not self._config.query.match(album) and not any(
        self._config.query.match(item) for item in album.items()
    ):
        continue
    dest_dir = self.album_destination(album)
    if not dest_dir:
        continue

    artpath = album.artpath and Path(str(album.artpath, "utf8"))
    if not artpath or not artpath.is_file():
        continue
```

**Review by geigerzaehler — COMMENTED** [review]
Thanks for sticking with this. I think we’re close for this to be done. Also could you please add an entry under a new section `## Unreleased` to `CHANGELOG.md`. And make sure to tag yourself as the contributor ;)

**Comment by geigerzaehler** [comment]
> The `unclosed file` exception occurred here again: https://github.com/geigerzaehler/beets-alternatives/actions/runs/19389877239/job/55482053515#step:8:78

You’re probably not seeing the issue locally because the ImageMagick backend is used on your machine. This patch should silence the issue:
```patch
diff --git a/pyproject.toml b/pyproject.toml
index 521eb2f..1f607a7 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -54,6 +54,7 @@ filterwarnings = [
   "error",
   "ignore:.*pkgutil.get_loader.*:DeprecationWarning:confuse.util",
   "ignore:.*beets.art.*:DeprecationWarning:beets",
+  "ignore::ResourceWarning:beets.util.artresizer",
 ]
 
 [tool.pyright]
```

File: beetsplug/alternatives.py [hunk 6]
```diff
@@ -410,6 +457,80 @@ def finalize_converted_item(item: Item, dest: Path):
...
+
+            if link:
+                self._log.debug(f"Linking art from {album.artpath} into {dest_dir}")
+                util.link(artpath, dest, replace=True)
+            else:
+                path = self.resize_art(artpath)
+                self._log.debug(f"Copying art from {path} into {dest_dir}")
+                util.copy(path, dest, replace=True)
+
+            print_(f"~{dest_dir}")
```
Comment (geigerzaehler) [inline beetsplug/alternatives.py]: Shouldn’t this by `f"~{dest}"`. If so make sure that `dest` remains a `Path` and is not changes to `bytes`. Other wise the output will look like
```plain
~b'/home/....'
```

File: beetsplug/alternatives.py [hunk 7]
```diff
@@ -410,6 +457,80 @@ def finalize_converted_item(item: Item, dest: Path):
...
+
+            dest_dir = self.album_destination(album)
+            if not dest_dir:
+                continue
+
+            artpath = album.artpath and Path(str(album.artpath, "utf8"))
+            if not artpath or not artpath.is_file():
+                continue
+
+            dest = album.art_destination(album.artpath, bytes(str(dest_dir), "utf8"))
```
Comment (geigerzaehler) [inline beetsplug/alternatives.py]:
```suggestion
            dest = album.art_destination(album.artpath, bytes(dest_dir))
```

File: test/cli_test.py [hunk 8]
```diff
@@ -200,6 +200,25 @@ def test_invalid_link_type(self):
...
+        )
+        album.set_art(self.IMAGE_FIXTURE1)
+        album.store()
+        self.runcli("alt", "update", "by-year")
+
+        external_album_path = tmp_path / "beets_lib" / "by-year" / "1990" / "Thriller"
+        external_art_path = external_album_path / "COVER.png"
+
+        # Symlink is created
+        assert external_art_path.is_symlink()
```
Comment (geigerzaehler) [inline test/cli_test.py]: Let’s also check that the symlink points to the album art

```suggestion
        assert_symlink(external_art_path, Path(str(album.artpath, "utf8")))
```

File: README.md [hunk 9]
```diff
@@ -271,6 +266,34 @@ following settings.
   **`formats`** is `link`, it sets the type of links to create. For
   differences between link types and examples see [Symlink Views](#symlink-views).
 
+* **`album_art_embed`** Embed album art into the media file. Default `yes`
+
+* **`album_art_copy`** Copy album art files into the collection. If
+  `formats: link` is used then album art is linked instead. Filename for
+  cover art is determined by the [art_filename][art_filename] option.
+
+* **`album_art_maxwidth`** Downscale the embedded album art to a width
```
Comment (geigerzaehler) [inline README.md]:
```suggestion
* **`album_art_maxwidth`** If set, embedded album art and album art files
  copied into the collection will be resized according to this value. The aspect
  ratio of the image will be preserved. This is comparable to the setting with the
  same name of the [convert plugin][convert plugin].
```

File: README.md [hunk 10]
```diff
@@ -271,6 +266,34 @@ following settings.
...
+  of maximum `album_art_maxwidth` pixels. The aspect ratio of the image
+  will be preserved. This is comparable to the setting with the same
+  name of the [convert plugin][convert plugin].
```
Comment (geigerzaehler) [inline README.md]:
```suggestion
* **`album_art_maxwidth`** If set, resize album art to this maximum width while
  preserving aspect ratio. Comparable to the [convert plugin][convert plugin]
  setting with the same name.
```

File: test/cli_test.py [hunk 11]
```diff
@@ -350,6 +369,146 @@ def test_unkown_collection(self):
...
+        self.runcli("alt", "update", "myexternal")
+        assert not external_art_path.is_file()
+
+        album.set_art(self.IMAGE_FIXTURE1)
+        album.store()
+        self.runcli("alt", "update", "myexternal")
+        assert_same_file_content(external_art_path, self.IMAGE_FIXTURE1)
+
+        # Update art file
+        album.set_art(self.IMAGE_FIXTURE2)
```
Comment (geigerzaehler) [inline test/cli_test.py]: I guess I was to optimistic about removing `touch_art`. There were some build failures

```suggestion
        album.set_art(self.IMAGE_FIXTURE2)
        touch_art(album.artpath, external_art_path)
```

File: beetsplug/alternatives.py [hunk 12]
```diff
@@ -410,6 +457,80 @@ def finalize_converted_item(item: Item, dest: Path):
...
+                new_format = {"jpeg": "jpg"}.get(new_format, new_format)
+
+                fname = dest.parent / dest.stem
+                dest = fname.with_suffix("." + new_format)
+
+            if dest.is_file() and not (artpath.stat().st_mtime > dest.stat().st_mtime):
+                continue
+
+            dest = bytes(str(dest), "utf8")
+            artpath = bytes(str(artpath), "utf8")
```
[the remark this answers is not in the document]
Comment (maxaudron) [reply beetsplug/alternatives.py]: I've removed the explicit "utf8" from all cases of encoding to bytes, but the bytes -> path conversion [here](https://github.com/geigerzaehler/beets-alternatives/pull/176/files#diff-6920da8b508c95e53faad52f1c70f966fa8b7742905897a1a2f3e71b8694ef40R464) in the tests causes File not found errors if it isn't specified. Is that expected behavior, or is there another way to do this?

[the remark this answers is not in the document]
Comment (maxaudron) [reply beetsplug/alternatives.py]: from a quick bit of research the correct way to do the str conversion there would be to use `os.fsdecode()`

[the remark this answers is not in the document]
Comment (geigerzaehler) [reply beetsplug/alternatives.py]: I understand. When we convert from `bytes` to `Path` we keep the current code. Only in the other direction can we apply `bytes()` directly.

**Comment by maxaudron** [comment]
should be good now

**Comment by geigerzaehler** [comment]
Great work @maxaudron. I’ll release a new version right away

**Comment by maxaudron** [comment]
Thank you! It was enjoyable working with you :)
